automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoSpec.scala

```diff
@@ -1,56 +1,195 @@
+    val cluster = Leonardo.cluster.create(googleProject, clusterName, clusterRequest)
+    clusterCheck(cluster, clusterRequest.labels, clusterName, Seq(ClusterStatus.Creating))
+
+    // verify with get()
+    clusterCheck(Leonardo.cluster.get(googleProject, clusterName), clusterRequest.labels, clusterName, Seq(ClusterStatus.Creating))
+
+    // wait for "Running" or error (fail fast)
+    val runningCluster = eventually {
+      clusterCheck(Leonardo.cluster.get(googleProject, clusterName), clusterRequest.labels, clusterName, Seq(ClusterStatus.Running, ClusterStatus.Error))
+    } (PatienceConfig(timeout = scaled(Span(8, Minutes)), interval = scaled(Span(20, Seconds))))
+
+    // now check that it didn't timeout or error
+    runningCluster.status shouldBe ClusterStatus.Running
+    runningCluster
+  }
+
+  // deletes a cluster and checks to see that it reaches the Deleted state
+  def deleteAndMonitor(googleProject: GoogleProject, clusterName: ClusterName): Unit = {
+    try {
+      Leonardo.cluster.delete(googleProject, clusterName) shouldBe
+        "The request has been accepted for processing, but the processing has not been completed."
+    } catch {
+      // OK if cluster not found / already deleted
+      case ae: APIException if ae.message.contains("\"statusCode\":404") => ()
+      case e: Exception => throw e
+    }
+
+    // wait until not found or in "Deleted" state
+    eventually {
+      val statusOpt = Leonardo.cluster.listIncludingDeleted().find(_.clusterName == clusterName).map(_.status)
+      statusOpt getOrElse ClusterStatus.Deleted shouldBe ClusterStatus.Deleted
+    } (PatienceConfig(timeout = scaled(Span(8, Minutes)), interval = scaled(Span(20, Seconds))))
+  }
+
+
+  // TODO: show diffs as screenshot or other test output?
+  def compareFilesExcludingIPs(left: File, right: File): Unit = {
+
+    def linesWithoutIPs(file: File) = {
+      import scala.collection.JavaConverters._
+      Files.readAllLines(left.toPath).asScala map { _.replaceAll("(\\d+.){3}\\d+", "<IP>") }
+    }
+
+    linesWithoutIPs(left) shouldEqual linesWithoutIPs(right)
+  }
+
+  // ------------------------------------------------------
+
+  def withNewCluster[T](googleProject: GoogleProject)(testCode: Cluster => T): T = {
+    val name = ClusterName(s"automation-test-a${Random.alphanumeric.take(10).mkString.toLowerCase}z")
+    val request = ClusterRequest(bucket, Map("foo" -> Random.alphanumeric.take(10).mkString))
+
+    val testResult: Try[T] = Try {
+      val cluster = createAndMonitor(googleProject, name, request)
+      testCode(cluster)
+    }
+
+    // delete before checking testCode status, which may throw
+    deleteAndMonitor(googleProject, name)
+    testResult.get
+  }
+
+  def withNotebooksListPage[T](cluster: Cluster)(testCode: NotebooksListPage => T)(implicit webDriver: WebDriver): T = {
+    // GAWB-2797: Leo might not be ready to proxy yet
+    import concurrent.duration._
+    Thread sleep 10.seconds.toMillis
```

Comment:
This is a workaround for a Leonardo bug: GAWB-2797

Yes normally we would want to fix the code rather than making allowances for it, but we are focusing on the basics before a deadline, alas

Reply:
My point was more that the sleep should be after something happens rather than before. This feels to me like paying the time penalty in the wrong place. Should this be in `withNewCluster` instead?

Reply:
I see what you mean.

Not sure that `withNewCluster` is right either.  We do have a cluster before this wait period.  But for some of these tests, we need a "Ready" cluster.  So I'll make that explicit with `withReadyCluster`